src/main/kotlin/com/openlattice/organizations/HazelcastOrganizationService.kt

```diff
@@ -285,44 +290,47 @@ class HazelcastOrganizationService(
         }
     }
 
+    /**
+     * Do not invoke this directly as it bypasses checks in [HazelcastOrganizationService.addMembers]
+     *
+     */
     private fun addMembers(
             orgAclKey: AclKey,
-            membersToAdd: Set<Principal>,
+            membersToAdd: Map<Principal, AclKey>,
             profiles: Map<Principal, Map<String, Set<String>>>
     ): Set<Principal> {
         require(orgAclKey.size == 1) { "Organization acl key should only be of length 1" }
-
-        val members = if (membersToAdd is Serializable) membersToAdd else membersToAdd.toSet()
-
-        val nonUserPrincipals = members.filter { it.type != PrincipalType.USER }
-        require(nonUserPrincipals.isEmpty()) { "Cannot add non-users principals $nonUserPrincipals to organization $orgAclKey" }
+        val members = membersToAdd.keys.toSet()
         val organizationId = orgAclKey[0]
-
-
         val newMembers: Set<Principal> = organizations.executeOnKey(organizationId, OrganizationEntryProcessor {
-            val newMembers = members.filter { member -> !it.members.contains(member) }.toSet()
+            val newMembers = members - it.members
             it.members.addAll(newMembers)
-            return@OrganizationEntryProcessor Result(newMembers, newMembers.isNotEmpty())
+            //We're always persisting here.
+            return@OrganizationEntryProcessor Result(newMembers)
         }) as Set<Principal>
 
-        if (newMembers.isNotEmpty()) {
-            grantOrganizationPrincipals(organizationId, newMembers, orgAclKey, profiles)
-        }
+        //Always trigger as this won't cause a write to organizations table.
```

Comment:
Not a huge deal, but it does always trigger a read from the organizations mapstore (even if there's no work to be done) -- maybe we should add a 
```
if (members.isEmpty()) {
  return
}
```
at the beginning of `grantOrganizationPrincipals`

Reply:
This seems to make more sense in the public addMembers call, since we shouldn't do any work at all if members is empty.

Optimization we could do in grantOrganizationPrincipals is push building out the list of roles to be granted to an entry processor, which is then returned the changes are completed. Also looking at it our e-mail processing is probably not RFC-compliant either.

Reply:
Cool, so you mean restore the old
```
if (newMembers.isNotEmpty()) {
  grantOrganizationPrincipals(...)
}
```
code? Doesn't seem to avoid any additional work but I'm good with that (although putting that logic in `grantOrganizationPrincipals` seems like a more thorough way to avoid doing the extra load. But again, just one hazelcast read so not a huge deal)

Doing that role work in an EP also sounds ideal if possible.

How does this interact with email processing?